ratatool-sampling/src/main/scala/com/spotify/ratatool/samplers/util/SamplerSCollectionFunctions.scala

```diff
@@ -0,0 +1,253 @@
+                                                        varByKey: SCollection[(U, Double)],
+                                                        probByKey: SCollection[(U, Double)],
+                                                        popPerKey: SideInput[Double])
+  : SCollection[(U, Double)] = {
+    s.map{ case (k, (_, d)) => (k, d) }
+      .hashJoin(varByKey)
+      .hashJoin(probByKey)
+      .filter{case (_, ((d, variance), prob)) =>
+        d < (prob + variance) && d >= (prob - variance)}
+      .map{case (k, ((d, _), _)) => (k, d)}
+      //TODO: Clean up magic number
+      .topByKey(1e8.toInt)(Ordering.by(identity[Double]).reverse)
+      .hashJoin(countsByKey)
+      .hashJoin(varByKey)
+      .hashJoin(probByKey)
+      .withSideInputs(popPerKey)
+      .map { case ((k, ((((itr, (lower, upper)), variance)), prob)), sic) =>
+        if (lower >= sic(popPerKey)) {
+          (k, prob - variance)
+        }
+        else if (upper < sic(popPerKey)) {
+          (k, prob + variance)
+        }
+        else {
+          val threshold = itr.drop(max(0, (sic(popPerKey) - lower).toInt)).headOption
+          (k, threshold.getOrElse(prob + variance))
+        }
+      }.toSCollection
+  }
+
+  implicit class RatatoolKVDSCollection[T: ClassTag, U: ClassTag](s: SCollection[(U, (T, Double))]){
+    //scalastyle:off cyclomatic.complexity
+    /**
+     * Performs higher precision sampling for a given distribution by assigning random probabilities
+     * to each element in SCollection[T] and then finding the Kth lowest, where K is the target
+     * population for a given strata or key.
+     *
+     * Currently variance is used to create cut-offs for upper and lower bounds to find K,
+     * but this can be improved inthe future.
+     */
+    def exactSampleDist(dist: SampleDistribution, keyFn: T => U, prob: Double): SCollection[T] = {
+      @transient lazy val logSerDe = LoggerFactory.getLogger(this.getClass)
+
+      val (sampled, sampledDiffs) = dist match {
+        case StratifiedDistribution =>
+          val targets = s.countByKey.map { case (k, c) => (k, (c * prob).toLong) }
+          val variance = prob * (1 - prob)
+          val counts = s.filter { case (_, (_, d)) => d < (prob + variance) }
+            .map { case (k, (_, d)) => (k, (if (d < prob - variance) 1L else 0L, 1L)) }
+            .sumByKey
+
+          val thresholdByKey = stratifiedThresholdByKey(s, counts, targets, variance, prob)
+
+          val sample = filterByThreshold(s, thresholdByKey)
+          val diffs = buildStratifiedDiffs(s.values.keys, sample, keyFn, prob)
+          (sample, diffs)
+
+        case UniformDistribution =>
+          val (popPerKey, probPerKey) = uniformParams(s.values.keys, keyFn, prob)
+          //TODO: Find better bounds than variance, potentially what Spark uses
+          val variancePerKey = probPerKey.map{case (k, f) => (k, f * (1 - f))}
```

Comment:
Better bounds will reduce the amount of elements we need in memory. Sparks bounds can be found here: https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/util/random/SamplingUtils.scala#L147

Reply:
ok I think this answers my question above, would be good to preserve this in a comment tho?